### PR TITLE
feat: implement animated features demo component

### PR DESCRIPTION
**Pull request by shadowusr**
### What's done?
This PR implements the animated features demo component. Feel free to explore how it looks on different screen sizes in storybook.

Notes:
- All icons were taken from https://uxwing.com/ and require no attribution
- Book cover images were taken from https://unsplash.com/ and require no attribution

### How to review?
Consider using Github Codespaces for easier navigation.

`src/components/FeaturesDemo/index.tsx` is a main file describing the whole demo timeline, it's a good place to start.

### CODE REVIEW

File: src/components/FeaturesDemo/index.tsx [hunk 1]
```diff
@@ -0,0 +1,726 @@
...
+
+    const consoleElement = (
+        <>
+            {consoleRefs.map(({ messages, ref }, index) => (
+                <span ref={ref} key={`${messages[0]}_${index}`}></span>
+            ))}
+        </>
+    );
+
+    const controller = new AbortController();
```
Comment (shadowusr) [inline src/components/FeaturesDemo/index.tsx:104]: Used to prevent potentially dangling async actions form corrupting the state

File: src/components/FeaturesDemo/index.tsx [hunk 2]
```diff
@@ -0,0 +1,726 @@
...
+
+        return () => {
+            store.reset();
+            controller.abort();
+        };
+    }, []);
+
+    const isHorizontalScreen = useScreenOrientation() === ScreenOrientation.Horizontal;
+
+    const transitions = useTransition(store.windows, {
```
Comment (shadowusr) [inline src/components/FeaturesDemo/index.tsx:644]: docs: https://www.react-spring.dev/docs/components/use-transition

In essence, you:
1. Define transitions using `useTransition` hook
2. Define how each item should be rendered
3. Use `animated.*` components and apply styles from `react-spring` to animate them

File: tailwind.config.ts [hunk 3]
```diff
@@ -1,9 +1,25 @@
...
+            mono: ["Roboto Mono", "monospace"],
+        },
         extend: {},
     },
+    safelist: [
+        // Classes used for code highlighting
+        "text-green-400",
+        "text-pink-500",
+        "text-violet-400",
+        "italic",
```
Comment (DudaGod) [inline tailwind.config.ts:22]: Why we should describe these classes here?

File: src/components/Collapsible/index.tsx [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    title?: ReactNode;
+    children?: ReactNode;
+    isCollapsed?: boolean;
+}
+
+export function Collapsible(props: CollapsibleProps) {
+    return (
+        <div>
+            <div>{props.title}</div>
+            {!props.isCollapsed && <div className="pl-3">{props.children}</div>}
```
Comment (DudaGod) [inline src/components/Collapsible/index.tsx:13]: `pl-3` - it is `padding-left: 3`?

File: src/components/FeaturesDemo/index.tsx [hunk 5]
```diff
@@ -0,0 +1,726 @@
...
+                consoleCleanupRef,
+            );
+
+            // eslint-disable-next-line @typescript-eslint/non-nullable-type-assertion-style
+            const searchInput = appBrowserPageRef.current
+                ?.getUnsafePageRef()
+                ?.current?.querySelector('input[data-role-id="search-input"]') as HTMLInputElement;
+            const offset = searchInput.offsetTop;
+            appBrowserPageRef.current?.setScrollTop(
+                offset - 36 - 40 /* height of search bar, height of header */,
```
Comment (DudaGod) [inline src/components/FeaturesDemo/index.tsx:272]: Can we use some constant values here for 36 and 40 and use them in css styles?

File: src/components/FeaturesDemo/index.tsx [hunk 6]
```diff
@@ -0,0 +1,726 @@
...
+        },
+        async function clickRunButton() {
+            // eslint-disable-next-line @typescript-eslint/non-nullable-type-assertion-style
+            const runButton = reportBrowserPageRef.current
+                ?.getUnsafePageRef()
+                ?.current?.querySelector('button[data-role-id="run-tests"]') as HTMLButtonElement;
+
+            await reportCursorRef.current?.setPosition(
+                runButton.offsetLeft + 25 - 6,
+                runButton.offsetTop + 1 + 40,
```
Comment (DudaGod) [inline src/components/FeaturesDemo/index.tsx:415]: What these magic values (25, 6, 1, 40) mean?

File: src/components/FeaturesDemo/index.tsx [hunk 7]
```diff
@@ -0,0 +1,726 @@
...
+            if (diffImgSize.bottom > pageHeight) {
+                reportBrowserPageRef.current?.setScrollTop(diffImg.offsetTop - 60);
+            }
+            await pause(PAUSE_MD, signal);
+
+            const cursor = reportCursorRef.current!;
+            await cursor.show();
+            const scrollOffset = reportBrowserPageRef.current!.getScrollTop();
+            await cursor.setPosition(
+                diffImg.offsetLeft + diffImgSize.width / 2 - 6,
```
Comment (DudaGod) [inline src/components/FeaturesDemo/index.tsx:555]: 6 - what this value mean?

File: src/components/FeaturesDemo/CodeEditor/styles.module.scss [hunk 8]
```diff
@@ -0,0 +1,90 @@
+.wrapper {
```
Comment (DudaGod) [inline src/components/FeaturesDemo/CodeEditor/styles.module.scss:1]: Why you choose scss instead stylus?

**Review by DudaGod — APPROVED** [review]
:fire:

File: src/components/FeaturesDemo/constants.ts [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+        ],
+    },
+    { id: CodeId.ReplTestEnd, messages: [`<br>  });<br>`] },
+    {
+        id: CodeId.BasicTestEnd,
+        messages: [
+            `<br>  it("should render product card", async ({browser}) => {
+    await browser.url("localhost:8000");
+
+    await browser.assertView("div[data-data-id='product']", "default-state");
```
Comment (DudaGod) [inline src/components/FeaturesDemo/constants.ts:89]: In this case first argument must be state and the second one selector.
But I suggest write it like this - `await browser.$("div[data-data-id='product']").assertView("default-state")`

File: tailwind.config.ts (hunk 3 above)
Comment (shadowusr) [reply tailwind.config.ts:22]: Hm, I've a comment above to explain that, guess it should be more comprehensive.

When building, Tailwind only includes classes you actually use in the output. These classes we use for code highlighting, but since code highlighting happens in runtime, Tailwind can't possibly know that we'll need those unless we explicitly list them here.

File: src/components/Collapsible/index.tsx (hunk 4 above)
Comment (shadowusr) [reply src/components/Collapsible/index.tsx:13]: `pl-3` is for `padding-left: 0.75rem; /* 12px */`. Tailwind has a set of pre-defined sizes that fit pretty well in my opinion the majority of use-cases. Tailwind devs maintain that it's a good idea to have a finite set of choices, because it eases decision-making as well as consistency of spacings across the app.

https://tailwindcss.com/docs/padding

File: src/components/FeaturesDemo/index.tsx [hunk 10]
```diff
@@ -0,0 +1,726 @@
...
+
+    const codeElement = (
+        <div
+            className="relative h-full w-full overflow-y-scroll scroll-smooth p-1"
+            ref={codeElementRef}
+        >
+            <div
+                className="absolute text-right text-blue-400 opacity-40"
+                dangerouslySetInnerHTML={{
+                    __html: Array.from(Array(50).keys()).join("<br>"),
```
[the remark this answers is not in the document]
Comment (shadowusr) [reply src/components/FeaturesDemo/index.tsx:79]: It's not a hack, that's how we generate line numbers for the editor. The resulting html is `1<br>2<br>3<br>...`.

File: src/components/FeaturesDemo/index.tsx [hunk 11]
```diff
@@ -0,0 +1,726 @@
...
+                        tabs: SAMPLE_TABS,
+                        fsTreeRoots: SAMPLE_FS_TREE_ROOTS,
+                        code: codeElement,
+                        console: consoleElement,
+                    },
+                },
+                signal,
+            );
+
+            await pause(PAUSE_SM, signal);
```
[the remark this answers is not in the document]
Comment (shadowusr) [reply src/components/FeaturesDemo/index.tsx:128]: By trial and error mostly. Their values are discussable if you feel like the animation is too fast or too slow

File: src/components/FeaturesDemo/store.ts [hunk 12]
```diff
@@ -0,0 +1,89 @@
+import { CodeEditorProps } from "@site/src/components/FeaturesDemo/CodeEditor";
+import { AppPageProps } from "@site/src/components/FeaturesDemo/WebBrowser/AppPage";
+import { ReportPageProps } from "@site/src/components/FeaturesDemo/WebBrowser/ReportPage";
+import { WebBrowserProps } from "@site/src/components/FeaturesDemo/WebBrowser";
+import { PartialDeep } from "@site/src/types";
+import { create } from "zustand";
```
[the remark this answers is not in the document]
Comment (shadowusr) [reply src/components/FeaturesDemo/store.ts:6]: We can discuss this. My motivation was the following:
- It'll probably the only place in the whole app where we'll use some kind of state management
- It's a landing, so the library should be very tiny in size
- zustand allows you to easily create convenient API for working with your store

TL;DR, it's simple, it's tiny, it fits the use case really well.

File: src/components/FeaturesDemo/index.tsx [hunk 13]
```diff
@@ -0,0 +1,726 @@
...
+        },
+        async function typeBasicTest() {
+            await typeCode(getById(codeRefs, CodeId.BasicTestBegin), codeCleanupRef);
+            await typeCode(getById(codeRefs, CodeId.BasicTestEnd), codeCleanupRef);
```
[the remark this answers is not in the document]
Comment (shadowusr) [reply src/components/FeaturesDemo/index.tsx:132]: That's because of how `Typed.js` works. As you can see, in our use case, we need the following features:
- type some code
- go back to the middle of the code
- remove some code and type another one (for fixing input selector, for instance)

However, `Typed.js` has only one feature: type some code into an element. It doesn't let you type in the middle or move the cursor in any way. So, in order to circumvent this constraint, we split the code into spans like this: `BasicTestBegin|ReplTestBegin|...|ReplTestEnd|BasicTestEnd`, then we can fill those spans in any order we want, giving the illusion of typing in the middle.

I've spent a while trying to come up with a better solution for this case, but it seems like this is our best bet. There's more versatile option out there — https://www.typeitjs.com/, but it requires a paid license. Implementing this typing experience from scratch is not trivial either, I've tried it too.

File: src/components/FeaturesDemo/index.tsx [hunk 14]
```diff
@@ -0,0 +1,726 @@
...
+                        console: consoleElement,
+                    },
+                },
+                signal,
+            );
+
+            await pause(PAUSE_SM, signal);
+        },
+        async function typeBasicTest() {
+            await typeCode(getById(codeRefs, CodeId.BasicTestBegin), codeCleanupRef);
```
[the remark this answers is not in the document]
Comment (shadowusr) [reply src/components/FeaturesDemo/index.tsx:131]: React hooks can only be used at the top level of component, we can't use them inside async steps

File: src/components/FeaturesDemo/CodeEditor/styles.module.scss (hunk 8 above)
Comment (shadowusr) [reply src/components/FeaturesDemo/CodeEditor/styles.module.scss:1]: Sass seems to be far more popular than stylus. IDK, I'm used to working with Sass and I think they all are mostly the same. Do you have any reasons to choose stylus.

Most importantly, I don't think we need sass much at all, because Tailwind covers almost all our styling needs.

File: src/components/FeaturesDemo/WebBrowser/ReportPage/AssertViewResult.tsx [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+                            <TickIcon className="mr-1 size-3" />
+                            Accept
+                        </ActionButton>
+                        <ActionButton className="mr-1">
+                            <SearchIcon className="size-3 sm:mr-1" />
+                            <span className="hidden sm:inline">Find same diffs</span>
+                        </ActionButton>
+                        <ActionButton className="mr-1">
+                            <FullscreenIcon className="size-3 sm:mr-1" />
+                            <span className="hidden sm:inline">Fast accept mode</span>
```
Comment (sipayRT) [inline src/components/FeaturesDemo/WebBrowser/ReportPage/AssertViewResult.tsx:47]: why are you using the `hidden` value here?

File: src/components/FeaturesDemo/WebBrowser/ReportPage/AssertViewResult.tsx [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+                    <div className="mr-2 basis-1/3">
+                        <div className="text-sm font-medium text-gray-500">Expected</div>
+                        <img src={data.expectedSrc} alt="expected" data-role-id="expected-image" />
+                    </div>
+                )}
+                <div className="mr-2 basis-1/3">
+                    <div className="text-sm font-medium text-gray-500">Actual</div>
+                    <img src={data.actualSrc} alt="actual" data-role-id="actual-image" />
+                </div>
+                {data.status === Status.Diff && (
```
Comment (sipayRT) [inline src/components/FeaturesDemo/WebBrowser/ReportPage/AssertViewResult.tsx:78]: imho, it would be better to split html with 2 states (diff|success). Now you have a lot of duplications of the same conditions in your code

File: src/components/FeaturesDemo/WebBrowser/ReportPage/CollapsibleTest.tsx [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+    const icons =
+        entryType === EntryType.TestSection ? [CopyIcon, RetryIcon] : [ViewIcon, ShareIcon];
+
+    return (
+        <Collapsible
+            title={
+                <div
+                    className={clsx("flex items-center", {
+                        "text-green-600": data.status === Status.Success,
+                        // "text-black": data.status === Status.NotLaunched,
```
Comment (sipayRT) [inline src/components/FeaturesDemo/WebBrowser/ReportPage/CollapsibleTest.tsx:37]: !

File: src/components/FeaturesDemo/WebBrowser/ReportPage/CollapsibleTest.tsx [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+                                el && (el.indeterminate = data.state === TestState.Indeterminate)
+                            }
+                            onChange={() => {
+                                /* empty handler just to prevent the input from becoming read-only */
+                            }}
+                            disabled={data.isDisabled}
+                        />
+                    )}
+                    {data.state === TestState.Loading && (
+                        <div className="absolute left-2">
```
Comment (sipayRT) [inline src/components/FeaturesDemo/WebBrowser/ReportPage/CollapsibleTest.tsx:57]: why do you need `absolute` here? And I can't find `relative` value for parent

File: src/components/FeaturesDemo/WebBrowser/ReportPage/Label.tsx [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+    value: string;
+    className?: string;
+}
+
+export function Label(props: LabelProps) {
+    return (
+        <div
+            className={clsx("flex flex-nowrap text-nowrap text-xs text-gray-500", props.className)}
+        >
+            <div className="inline px-1 py-0.5 font-semibold">{props.title}</div>
```
Comment (sipayRT) [inline src/components/FeaturesDemo/WebBrowser/ReportPage/Label.tsx:15]: if you need `inline` here, use `span` for it instead of `div`

File: src/components/FeaturesDemo/WebBrowser/ReportPage/constants.ts [hunk 20]
```diff
@@ -0,0 +1,149 @@
...
+                        isCollapsed: true,
+                        isDisabled: false,
+                        state: TestState.Unchecked,
+                        status: Status.NotLaunched as Status,
+                        selectedAttempt: 0,
+                        attempts: [] as TestAttemptWithState[],
+                    },
+                ],
+            },
+            /*{
```
Comment (sipayRT) [inline src/components/FeaturesDemo/WebBrowser/ReportPage/constants.ts:75]: !

File: src/components/FeaturesDemo/WebBrowser/ReportPage/styles.module.scss [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+        transform: scale(1);
+    }
+}
+
+:global {
+    .diff-bubble.animated {
+        & :local {
+            animation-name: diff-bubble-appear;
+        }
+        /*animation-delay: 2s;*/
```
Comment (sipayRT) [inline src/components/FeaturesDemo/WebBrowser/ReportPage/styles.module.scss:22]: !

File: src/components/FeaturesDemo/WebBrowser/ReportPage/types.ts [hunk 22]
```diff
@@ -0,0 +1,88 @@
...
+    status: Status.Diff;
+    expectedSrc: string;
+    actualSrc: string;
+    diffSrc: string;
+    diffBubbles: { x: string; y: string }[];
+}
+
+export type AssertViewResult = AssertViewSuccess | AssertViewDiff;
+
+// export interface TestBranchItem {
```
Comment (sipayRT) [inline src/components/FeaturesDemo/WebBrowser/ReportPage/types.ts:31]: !

File: src/components/FeaturesDemo/WebBrowser/ReportPage/types.ts [hunk 23]
```diff
@@ -0,0 +1,88 @@
...
+    assertViewResults?: AssertViewResult[];
+}
+
+export interface TestResult {
+    title: string;
+    status: Status;
+    attempts: TestAttempt[];
+}
+
+// state ===
```
Comment (sipayRT) [inline src/components/FeaturesDemo/WebBrowser/ReportPage/types.ts:55]: !

File: src/components/FeaturesDemo/WebBrowser/index.tsx [hunk 24]
```diff
@@ -0,0 +1,88 @@
...
+            return pageRef.current?.scrollTop ?? 0;
+        },
+        getUnsafePageRef: (): RefObject<HTMLDivElement> => {
+            return pageRef;
+        },
+    }));
+
+    return (
+        <div className={clsx("min-w-0", props.className)} style={props.style}>
+            <div className={clsx(styles.browser, "relative")}>
```
Comment (sipayRT) [inline src/components/FeaturesDemo/WebBrowser/index.tsx:55]: why do you need `relative` here?

File: src/components/FeaturesDemo/AnimatedCursor.tsx [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+                        { scaleX: 1, scaleY: 1 },
+                    ],
+                }),
+            );
+        },
+        async setPosition(x, y): Promise<void> {
+            await Promise.all(
+                api.start({
+                    config: {
+                        /*duration: 500,*/
```
Comment (sipayRT) [inline src/components/FeaturesDemo/AnimatedCursor.tsx:43]: !

File: src/components/FeaturesDemo/constants.ts [hunk 26]
```diff
@@ -0,0 +1,164 @@
...
+        messages: ["`<br><span class='text-green-400'>< Success</span>`"],
+    },
+    { id: ConsoleId.SendHelloSearch, messages: ['`<br>> browser.$("HELLO");`'] },
+    {
+        id: ConsoleId.HelloSearchFailure,
+        messages: ["`<br><span class='text-red-500'>< Error: \"HELLO\" element not found</span>`"],
+    },
+    {
+        id: ConsoleId.SendInputSearch,
+        messages: ["`<br>> browser.$(\"input[data-data-id='search']\");`"],
```
[the remark this answers is not in the document]
Comment (sipayRT) [reply src/components/FeaturesDemo/constants.ts:131]: `.input_type_search` решает эту проблему ;)
как минимум, читается это проще

File: src/components/FeaturesDemo/index.tsx [hunk 27]
```diff
@@ -0,0 +1,726 @@
...
+    const codeElementRef = useRef<HTMLDivElement>(null);
+    const codeCleanupRef = useRef<() => void>(null);
+
+    const codeElement = (
+        <div
+            className="relative h-full w-full overflow-y-scroll scroll-smooth p-1"
+            ref={codeElementRef}
+        >
+            <div
+                className="absolute text-right text-blue-400 opacity-40"
```
Comment (sipayRT) [inline src/components/FeaturesDemo/index.tsx:77]: `absolute`... again... instead of simple `display: inline-block`

File: src/components/FeaturesDemo/index.tsx [hunk 28]
```diff
@@ -0,0 +1,726 @@
...
+            {consoleRefs.map(({ messages, ref }, index) => (
+                <span ref={ref} key={`${messages[0]}_${index}`}></span>
+            ))}
+        </>
+    );
+
+    const controller = new AbortController();
+    const signal = controller.signal;
+
+    const steps: (() => void | Promise<void>)[] = [
```
Comment (sipayRT) [inline src/components/FeaturesDemo/index.tsx:107]: imho, it would be better to move stept to the separate file(files) for more readability

File: src/components/FeaturesDemo/utils.ts [hunk 29]
```diff
@@ -0,0 +1,124 @@
...
+) => {
+    return type({ ...data, typedJsOptions: { cursorChar: "▊" } }, cleanupRef);
+};
+
+export const scrollToLineIfNeeded = (
+    lineRef: RefObject<HTMLSpanElement>,
+    containerRef: RefObject<HTMLDivElement>,
+) => {
+    const offsetTop =
+        lineRef.current!.offsetTop - containerRef.current!.parentElement!.clientHeight + 80;
```
Comment (sipayRT) [inline src/components/FeaturesDemo/utils.ts:90]: 80?

**Review by sipayRT — APPROVED** [review]

File: src/components/FeaturesDemo/constants.ts (hunk 26 above)
[the remark this answers is not in the document]
Comment (shadowusr) [reply src/components/FeaturesDemo/constants.ts:131]: Вроде бы использование конкретных testid'ов в тестах же является common-, если не best-practice. Я вижу такие преимущества использования разметки data-testid над селекторами:
- все data-testid можно вырезать при сборке бандла в прод, а селекторы нет
- селекторы относятся к стилям, если захочется рефакторить стили, поменяются селекторы и все тесты надо будет тоже чинить

File: src/components/FeaturesDemo/index.tsx (hunk 6 above)
Comment (shadowusr) [reply src/components/FeaturesDemo/index.tsx:415]: 40 is a constant, 25 is just a small offset — left a comment there.

File: src/components/FeaturesDemo/index.tsx (hunk 5 above)
Comment (shadowusr) [reply src/components/FeaturesDemo/index.tsx:272]: replaced with constants.

File: src/components/FeaturesDemo/index.tsx (hunk 7 above)
Comment (shadowusr) [reply src/components/FeaturesDemo/index.tsx:555]: got rid of it.

File: .storybook/main.ts [hunk 30]
```diff
@@ -43,25 +70,16 @@ const config: StorybookConfig = {
                 rules: [
                     ...(config.module?.rules as Record<string, any>[]),
```
[the remark this answers is not in the document]
Comment (shadowusr) [reply .storybook/main.ts:71]: Here, we extend storybook's webpack config with some custom rules. By adding `...config.module.rules` we make sure to include original rules, we don't want to throw them away when adding ours.

Docs: https://storybook.js.org/docs/api/main-config-webpack-final

File: eslint.config.mjs [hunk 31]
```diff
@@ -47,7 +48,11 @@ export default tseslint.config(
     reactRecommended,
     reactJsxRuntime,
     {
-        rules: geminiTesting.rules,
+        rules: _.omit(geminiTesting.rules, [
+            // Rules that conflict with typescript-eslint.
+            "no-undef",
+            "no-unused-vars",
```
[the remark this answers is not in the document]
Comment (shadowusr) [reply eslint.config.mjs:54]: Because it conflicts with typescript-eslint, as stated in the comment above. It provides its own rule, `@typescript-eslint/no-unusd-vars`, because the original one doesn't work well with TS code.

File: static/img/landing/bookstore/book-1.jpg
[the remark this answers is not in the document]
Comment (shadowusr) [reply static/img/landing/bookstore/book-1.jpg:1]: What's wrong with jpg? I'm trying to squeeze assets as tightly as I can so that landing page opens up fast enough even on slow internet. Pics are not of great quality, but good enough for demo purposes, IMO

File: src/components/Collapsible/index.tsx [hunk 32]
```diff
@@ -0,0 +1,16 @@
...
+
+interface CollapsibleProps {
+    title?: ReactNode;
+    children?: ReactNode;
+    isCollapsed?: boolean;
+}
+
+export function Collapsible(props: CollapsibleProps) {
+    return (
+        <div>
```
[the remark this answers is not in the document]
Comment (shadowusr) [reply src/components/Collapsible/index.tsx:11]: OK, added unused `collapsible` class for the sake of better devtools experience and corresponding class names for other components as well.

File: src/components/FeaturesDemo/CodeEditor/index.tsx [hunk 33]
```diff
@@ -0,0 +1,179 @@
...
+    title: string;
+    icon?: ReactElement;
+    isActive?: boolean;
+}
+
+function Tab(props: TabProps) {
+    return (
+        <div className={clsx(styles["tab-item"], "relative ml-5 flex items-center text-nowrap")}>
+            <div className={clsx({ "text-indigo-200": props.isActive })}>{props.title}</div>
+            <div className={clsx("ml-5 size-3", { "fill-indigo-200": props.isActive })}>
```
[the remark this answers is not in the document]
Comment (shadowusr) [reply src/components/FeaturesDemo/CodeEditor/index.tsx:26]: We do need it — in case the icon is absent, we use `<CloseIcon />`

File: src/components/FeaturesDemo/WebBrowser/ReportPage/AssertViewResult.tsx (hunk 15 above)
Comment (shadowusr) [reply src/components/FeaturesDemo/WebBrowser/ReportPage/AssertViewResult.tsx:47]: In order to free up as much space as possible on small phone screens. `hidden sm:inline` means that the block won't visible on mobile screens, but on larger ones will be inline.

File: src/components/FeaturesDemo/WebBrowser/AppPage/index.tsx [hunk 34]
```diff
@@ -0,0 +1,123 @@
...
+                        <ProductCard
+                            key={index}
+                            price={product.price}
+                            title={product.title}
+                            photoSrc={product.photoSrc}
+                            className="mb-4"
+                        />
+                    ))}
+            </div>
+            <div className="relative flex h-96 justify-between bg-gray-500 p-6 bg-blend-hard-light">
```
[the remark this answers is not in the document]
Comment (shadowusr) [reply src/components/FeaturesDemo/WebBrowser/AppPage/index.tsx:104]: Because below we have an absolutely positioned element — the `STYLE STYLE STYLE` text in background.

File: src/components/FeaturesDemo/WebBrowser/AppPage/index.tsx [hunk 35]
```diff
@@ -0,0 +1,123 @@
...
+    }));
+
+    return (
+        <div className="w-full min-w-0 bg-white">
+            <div className="sticky top-0 z-20 flex w-full flex-nowrap justify-between bg-white bg-opacity-35 p-1 align-middle backdrop-blur">
+                <div className="flex items-baseline">
+                    <span className="whitespace-nowrap pl-1 text-lg font-medium italic">
+                        Awesome store.
+                    </span>
+                    <span className="ml-3 hidden sm:block">Home</span>
```
[the remark this answers is not in the document]
Comment (shadowusr) [reply src/components/FeaturesDemo/WebBrowser/AppPage/index.tsx:46]: It works just fine as far as I can tell. `hidden sm:block` means that the block isn't visible on mobile screens, but appears on screens of larger size. If you open this page and check how it behaves on different screen sizes, you'll the difference.

Replaces all spans with divs.

File: src/components/FeaturesDemo/CodeEditor/index.tsx [hunk 36]
```diff
@@ -0,0 +1,179 @@
...
+                    <AngleIcon
+                        className={clsx(
+                            "absolute -left-5 top-1/2 flex size-3 -translate-y-1/2 items-center",
+                            { "rotate-90": !props.isCollapsed },
+                        )}
+                    />
+                )}
+            </div>
+            {props.isSelected && (
+                <div className="absolute left-0 z-0 h-6 w-full bg-indigo-900"></div>
```
[the remark this answers is not in the document]
Comment (shadowusr) [reply src/components/FeaturesDemo/CodeEditor/index.tsx:83]: Because I want the selection background width to be 100% of the whole fs tree column. However, each element takes only part of the column's width, so I can't just use background-color and need this artificial element for that.

File: src/components/FeaturesDemo/CodeEditor/index.tsx [hunk 37]
```diff
@@ -0,0 +1,179 @@
...
+        <div
+            className={clsx("ml-5 flex items-center", {
+                "fill-indigo-400 text-indigo-400": props.isModified,
+            })}
+        >
+            <div className="relative z-10">
+                {props.type === FsTreeItemType.Dir && (
+                    <AngleIcon
+                        className={clsx(
+                            "absolute -left-5 top-1/2 flex size-3 -translate-y-1/2 items-center",
```
[the remark this answers is not in the document]
Comment (shadowusr) [reply src/components/FeaturesDemo/CodeEditor/index.tsx:76]: vertical centering of absolutely positioned block, just as you'd write in plain css:
```css
top: 50%;
transform: translateY(-50%);
```

File: src/components/FeaturesDemo/WebBrowser/ReportPage/Label.tsx (hunk 19 above)
Comment (shadowusr) [reply src/components/FeaturesDemo/WebBrowser/ReportPage/Label.tsx:15]: idk, why rely on default browser styling if we can use divs everywhere and style them as we want. but alright, replaced with spans

File: src/components/FeaturesDemo/WebBrowser/index.tsx (hunk 24 above)
Comment (shadowusr) [reply src/components/FeaturesDemo/WebBrowser/index.tsx:55]: So that page content could have absolutely positioned elements. For instance, we use it to have sticky header on app page or when using virtual cursor to click on buttons.

File: src/components/FeaturesDemo/WebBrowser/ReportPage/CollapsibleTest.tsx (hunk 18 above)
Comment (shadowusr) [reply src/components/FeaturesDemo/WebBrowser/ReportPage/CollapsibleTest.tsx:57]: because I want checkboxes and loading spinners to be posiitoned at the very left of the page. `relative` is set on browser window.

File: src/components/FeaturesDemo/WebBrowser/AppPage/index.tsx [hunk 38]
```diff
@@ -0,0 +1,123 @@
...
+                </div>
+            </div>
+            <div className="p-3">
+                <input
+                    ref={inputRef}
+                    placeholder="Search"
+                    data-role-id="search-input"
+                    className="w-full rounded-full border border-gray-300 px-3 py-0.5 focus:outline-blue-300"
+                    value={props.searchText}
+                    readOnly={true}
```
[the remark this answers is not in the document]
Comment (shadowusr) [reply src/components/FeaturesDemo/WebBrowser/AppPage/index.tsx:86]: React complains that the input should be either readonly or have a handler. This component is used for demo purposes only, so no need for additional handlers and we can safely mark it as readonly.

File: src/components/FeaturesDemo/index.tsx (hunk 28 above)
Comment (shadowusr) [reply src/components/FeaturesDemo/index.tsx:107]: yeah, this thought has crossed my mind, however A LOT of things (many refs, signals, store data, etc.) are used inside steps (via closures), if we move them outside, we'll severely worsen readability in that aspect, so in the end I decided against it.

Otherwise we'll need to construct a function which takes 20+ params (even if packed into an object) and write corresponding types for them all.

Do you think it's worth it?

File: src/components/FeaturesDemo/CodeEditor/index.tsx [hunk 39]
```diff
@@ -0,0 +1,179 @@
...
+    code?: ReactNode;
+    console?: ReactNode;
+    status?: string;
+    tabs: TabProps[];
+    fsTreeRoots: FsTreeItem[];
+}
+
+export function CodeEditor(props: CodeEditorProps): JSX.Element {
+    return (
+        <div className={clsx(styles.wrapper, props.className)} style={props.style}>
```
[the remark this answers is not in the document]
Comment (shadowusr) [reply src/components/FeaturesDemo/CodeEditor/index.tsx:113]: Added class name with component name to most components.
